Sources/swift-format/Run.swift

```diff
@@ -29,38 +29,34 @@ import TSCBasic
 /// - Returns: Zero if there were no lint errors, otherwise a non-zero number.
 func lintMain(
   configuration: Configuration, sourceFile: FileHandle, assumingFilename: String?,
-  debugOptions: DebugOptions, diagnosticEngine: DiagnosticEngine
-) -> Int {
+  debugOptions: DebugOptions
+) throws {
+  let diagnosticEngine = makeDiagnosticEngine()
   let linter = SwiftLinter(configuration: configuration, diagnosticEngine: diagnosticEngine)
   linter.debugOptions = debugOptions
   let assumingFileURL = URL(fileURLWithPath: assumingFilename ?? "<stdin>")
 
   guard let source = readSource(from: sourceFile) else {
-    diagnosticEngine.diagnose(
-      Diagnostic.Message(.error, "Unable to read source for linting from \(assumingFileURL.path)."))
-    return 1
+    throw FormatError.readSource(path: assumingFileURL.path)
```

Comment:
In https://github.com/apple/swift-format/pull/144, @dylansturg refactored `lintMain` and `formatMain` a bit so that the error messages here were sent through the diagnostic engine (instead of being written directly to `stderr`)—this unified all possible error messages that the tool might emit after argument parsing. For example, we could add a flag later to replace the diagnostic consumer with one that serializes to JSON instead of printing them to `stderr`, and file I/O errors would be handled seamlessly.

Changing these to throw errors here would revert that back to having two different kinds of behavior for different kinds of errors. Do we need this change?

Reply:
I'll definitely revert that — it looked like I was replicating the current usage, but of course coalescing errors to look forward at using other outputs makes tons of sense.